File: aws-globalaccelerator-endpointgroup/src/test/java/software/amazon/globalaccelerator/endpointgroup/DeleteHandlerTest.java
package software.amazon.globalaccelerator.endpointgroup;

import com.amazonaws.services.globalaccelerator.model.*;
import lombok.val;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.Mock;
import org.mockito.junit.jupiter.MockitoExtension;
import software.amazon.cloudformation.proxy.AmazonWebServicesClientProxy;
import software.amazon.cloudformation.proxy.Logger;
import software.amazon.cloudformation.proxy.OperationStatus;
import software.amazon.cloudformation.proxy.ResourceHandlerRequest;

import java.util.ArrayList;

import static org.assertj.core.api.Assertions.assertThat;
import static org.assertj.core.api.Assertions.assertThatExceptionOfType;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.Mockito.*;

@ExtendWith(MockitoExtension.class)
public class DeleteHandlerTest {

    @Mock
    private AmazonWebServicesClientProxy proxy;

    @Mock
    private Logger logger;

    @BeforeEach
    public void setup() {
        proxy = mock(AmazonWebServicesClientProxy.class);
        logger = mock(Logger.class);
    }

    @Test
    public void handleRequest_ResourceStillExists_DeletesAndReturnsInProgress() {
        // return an accelerator that is IN_PROGRESS
        val describeAcceleratorResult = new DescribeAcceleratorResult()
                .withAccelerator(new Accelerator()
                        .withStatus(AcceleratorStatus.IN_PROGRESS));
        doReturn(describeAcceleratorResult).when(proxy).injectCredentialsAndInvoke(any(DescribeAcceleratorRequest.class), any());

        // we expect delete to be called
        doReturn(new DeleteEndpointGroupResult()).when(proxy).injectCredentialsAndInvoke(any(DeleteEndpointGroupRequest.class), any());

        // the endpoint group describe request
        val describeEndpointGroupResult = new DescribeEndpointGroupResult()
                .withEndpointGroup(new EndpointGroup()
                        .withEndpointGroupArn("ENDPOINT_GROUP_ARN"));
        doReturn(describeEndpointGroupResult).when(proxy).injectCredentialsAndInvoke(any(DescribeEndpointGroupRequest.class), any());

        // create the that will go to our handler
        val model = ResourceModel.builder()
                .endpointGroupRegion("us-west-2")
                .listenerArn("arn:aws:globalaccelerator::474880776455:accelerator/abcd1234/listener/12341234")
                .endpointGroupArn("ENDPOINT_GROUP_ARN")
                .healthCheckPort(-1)
                .thresholdCount(3)
                .build();

        val request = ResourceHandlerRequest.<ResourceModel>builder().desiredResourceState(model).build();
        val handler = new DeleteHandler();
        val response = handler.handleRequest(proxy, request, null, logger);

        assertThat(response).isNotNull();
        assertThat(response.getStatus()).isEqualTo(OperationStatus.IN_PROGRESS);
        assertThat(response.getCallbackDelaySeconds()).isEqualTo(1);
        assertThat(response.getCallbackContext()).isNotNull();
        assertThat(response.getResourceModels()).isNull();
        assertThat(response.getMessage()).isNull();

        // want to confirm model remains the same
        assertThat(response.getResourceModel()).isEqualTo(model);
    }

    @Test
    public void handleRequest_ResourceDoesNotExists_AcceleratorInProgress_ReturnsInProgress() {
        // return an accelerator that is IN_PROGRESS
        val describeAcceleratorResult = new DescribeAcceleratorResult()
                .withAccelerator(new Accelerator()
                        .withStatus(AcceleratorStatus.IN_PROGRESS));
        doReturn(describeAcceleratorResult).when(proxy).injectCredentialsAndInvoke(any(DescribeAcceleratorRequest.class), any());

        // the endpoint group describe request will return not found
        doThrow(new EndpointGroupNotFoundException("NOT FOUND"))
                .when(proxy).injectCredentialsAndInvoke(any(DescribeEndpointGroupRequest.class), any());

        // create the that will go to our handler
        val model = ResourceModel.builder()
                .endpointGroupRegion("us-west-2")
                .listenerArn("arn:aws:globalaccelerator::474880776455:accelerator/abcd1234/listener/12341234")
                .endpointGroupArn("ENDPOINT_GROUP_ARN")
                .healthCheckPort(-1)
                .thresholdCount(3)
                .build();

        val request = ResourceHandlerRequest.<ResourceModel>builder().desiredResourceState(model).build();
        val handler = new DeleteHandler();
        val context = CallbackContext.builder().stabilizationRetriesRemaining(2).build();
        val response = handler.handleRequest(proxy, request, context, logger);

        assertThat(response).isNotNull();
        assertThat(response.getStatus()).isEqualTo(OperationStatus.IN_PROGRESS);
        assertThat(response.getCallbackDelaySeconds()).isEqualTo(1);
        assertThat(response.getCallbackContext()).isNotNull();
        assertThat(response.getResourceModels()).isNull();
        assertThat(response.getMessage()).isNull();

        // want to confirm model remains the same
        assertThat(response.getResourceModel()).isEqualTo(model);
    }

    @Test
    public void handleRequest_ResourceDoesNotExist_AcceleratorDeployed_ReturnsSuccess() {
        // return an accelerator that is IN_PROGRESS
        val describeAcceleratorResult = new DescribeAcceleratorResult()
                .withAccelerator(new Accelerator()
                        .withStatus(AcceleratorStatus.DEPLOYED));
        doReturn(describeAcceleratorResult).when(proxy).injectCredentialsAndInvoke(any(DescribeAcceleratorRequest.class), any());

        // the endpoint group describe request will return not found
        doThrow(new EndpointGroupNotFoundException("NOT FOUND"))
                .when(proxy).injectCredentialsAndInvoke(any(DescribeEndpointGroupRequest.class), any());

        // create the that will go to our handler
        val model = ResourceModel.builder()
                .endpointGroupRegion("us-west-2")
                .listenerArn("arn:aws:globalaccelerator::474880776455:accelerator/abcd1234/listener/12341234")
                .endpointGroupArn("ENDPOINT_GROUP_ARN")
                .healthCheckPort(-1)
                .thresholdCount(3)
                .build();

        val request = ResourceHandlerRequest.<ResourceModel>builder().desiredResourceState(model).build();
        val handler = new DeleteHandler();
        val context = CallbackContext.builder().stabilizationRetriesRemaining(2).build();
        val response = handler.handleRequest(proxy, request, context, logger);

        assertThat(response).isNotNull();
        assertThat(response.getStatus()).isEqualTo(OperationStatus.SUCCESS);
        assertThat(response.getCallbackDelaySeconds()).isEqualTo(0);
        assertThat(response.getCallbackContext()).isNull();
        assertThat(response.getResourceModels()).isNull();
        assertThat(response.getMessage()).isNull();

        // want to confirm model remains the same
        assertThat(response.getResourceModel()).isEqualTo(model);
    }

    @Test
    public void handleRequest_ResourceDoesNotExist_ThresholdTimeExceeded() {
        // the endpoint group describe request will return not found
        doThrow(new EndpointGroupNotFoundException("NOT FOUND"))
                .when(proxy).injectCredentialsAndInvoke(any(DescribeEndpointGroupRequest.class), any());

        // Create the model we will provide to our handler
        val endpointConfigurations = new ArrayList<EndpointConfiguration>();
        endpointConfigurations.add(EndpointConfiguration.builder().endpointId("EPID1").build());
        endpointConfigurations.add(EndpointConfiguration.builder().endpointId("EPID2").build());
        val model = ResourceModel.builder()
                .endpointGroupRegion("us-west-2")
                .listenerArn("arn:aws:globalaccelerator::474880776455:accelerator/abcd1234/listener/12341234")
                .endpointGroupArn("ENDPOINT_GROUP_ARN")
                .healthCheckPort(20)
                .thresholdCount(3)
                .trafficDialPercentage(100)
                .healthCheckPath("/HEALTH")
                .endpointConfigurations(endpointConfigurations)
                .build();

        // call the handler
        val request = ResourceHandlerRequest.<ResourceModel>builder().desiredResourceState(model).build();
        val callbackContext = CallbackContext.builder().stabilizationRetriesRemaining(0).build();
        val handler = new DeleteHandler();

        assertThatExceptionOfType(RuntimeException.class).isThrownBy(() -> handler.handleRequest(proxy, request, callbackContext, logger))
                .withMessageMatching("Timed out waiting for endpoint group to be deployed.");
    }
}

File: aws-globalaccelerator-endpointgroup/src/test/kotlin/software/amazon/globalaccelerator/endpointgroup/DeleteHandlerTest.kt
package software.amazon.globalaccelerator.endpointgroup

import com.amazonaws.services.globalaccelerator.model.Accelerator
import com.amazonaws.services.globalaccelerator.model.AcceleratorNotFoundException
import com.amazonaws.services.globalaccelerator.model.AcceleratorStatus
import com.amazonaws.services.globalaccelerator.model.DeleteEndpointGroupRequest
import com.amazonaws.services.globalaccelerator.model.DeleteEndpointGroupResult
import com.amazonaws.services.globalaccelerator.model.DescribeAcceleratorRequest
import com.amazonaws.services.globalaccelerator.model.DescribeAcceleratorResult
import com.amazonaws.services.globalaccelerator.model.DescribeEndpointGroupRequest
import com.amazonaws.services.globalaccelerator.model.DescribeEndpointGroupResult
import com.amazonaws.services.globalaccelerator.model.EndpointGroup
import com.amazonaws.services.globalaccelerator.model.EndpointGroupNotFoundException
import io.mockk.MockKAnnotations
import io.mockk.every
import io.mockk.impl.annotations.MockK
import io.mockk.junit5.MockKExtension
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Assertions.assertNotNull
import org.junit.jupiter.api.Assertions.assertNull
import org.junit.jupiter.api.Assertions.assertTrue
import org.junit.jupiter.api.Assertions.assertThrows
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.extension.ExtendWith
import software.amazon.cloudformation.proxy.AmazonWebServicesClientProxy
import software.amazon.cloudformation.proxy.HandlerErrorCode
import software.amazon.cloudformation.proxy.Logger
import software.amazon.cloudformation.proxy.OperationStatus
import software.amazon.cloudformation.proxy.ResourceHandlerRequest
import java.util.ArrayList
import java.util.function.Function

@ExtendWith(MockKExtension::class)
class DeleteHandlerTest {
    @MockK
    lateinit var proxy: AmazonWebServicesClientProxy

    @MockK(relaxed = true)
    lateinit var logger: Logger

    @BeforeEach
    fun setup() = MockKAnnotations.init(this)

    private val endpointGroupRegion = "us-west-2"
    private val listenerArn = "arn:aws:globalaccelerator::444607872184:accelerator/88127aa5-01d8-484c-80a0-349daaefce1d/listener/ee7358c2"
    private val endpointGroupArn = "arn:aws:globalaccelerator::444607872184:accelerator/88127aa5-01d8-484c-80a0-349daaefce1d/listener/ee7358c2/endpoint-group/de69a4b45005"


    @Test
    fun handleRequest_ResourceStillExists_DeletesAndReturnsInProgress() {
        val describeEndpointGroupResult = DescribeEndpointGroupResult()
                .withEndpointGroup(EndpointGroup()
                        .withEndpointGroupArn(endpointGroupArn))
        every { proxy.injectCredentialsAndInvoke(ofType(), ofType<ProxyDescribeEndpointGroup>()) } returns describeEndpointGroupResult

        every { proxy.injectCredentialsAndInvoke(ofType(), ofType<ProxyDeleteEndpointGroup>()) } returns DeleteEndpointGroupResult()

        val model = ResourceModel.builder()
                .endpointGroupRegion(endpointGroupRegion)
                .listenerArn(listenerArn)
                .endpointGroupArn(endpointGroupArn)
                .healthCheckPort(-1)
                .thresholdCount(3)
                .build()

        val request = ResourceHandlerRequest.builder<ResourceModel>().desiredResourceState(model).build()
        val response = DeleteHandler().handleRequest(proxy, request, null, logger)

        assertNotNull(response)
        assertEquals(OperationStatus.IN_PROGRESS, response.status)
        assertEquals(0, response.callbackDelaySeconds)
        assertNotNull(response.callbackContext)
        assertNull(response.message)
        assertNull(response.resourceModels)
        assertEquals(model, response.resourceModel)
    }

    @Test
    fun handleRequest_ResourceDoesNotExists_AcceleratorInProgress_ReturnsInProgress() {
        val describeAcceleratorResult = DescribeAcceleratorResult()
                .withAccelerator(Accelerator()
                        .withStatus(AcceleratorStatus.IN_PROGRESS))
        every { proxy.injectCredentialsAndInvoke(ofType(), ofType<ProxyDescribeAccelerator>()) } returns describeAcceleratorResult

        val model = ResourceModel.builder()
                .endpointGroupRegion(endpointGroupRegion)
                .listenerArn(listenerArn)
                .endpointGroupArn(endpointGroupArn)
                .healthCheckPort(-1)
                .thresholdCount(3)
                .build()
        val request = ResourceHandlerRequest.builder<ResourceModel>().desiredResourceState(model).build()
        val context = CallbackContext(2, pendingStabilization = true)

        val response = DeleteHandler().handleRequest(proxy, request, context, logger)

        assertNotNull(response)
        assertEquals(OperationStatus.IN_PROGRESS, response.status)
        assertEquals(1, response.callbackDelaySeconds)
        assertNotNull(response.callbackContext)
        assertNull(response.message)
        assertNull(response.resourceModels)
        assertEquals(model, response.resourceModel)
    }

    @Test
    fun handleRequest_ResourceDoesNotExist_AcceleratorDeployed_ReturnsSuccess() {
        val model = ResourceModel.builder()
                .endpointGroupRegion(endpointGroupRegion)
                .listenerArn(listenerArn)
                .endpointGroupArn(endpointGroupArn)
                .healthCheckPort(-1)
                .thresholdCount(3)
                .build()

        every { proxy.injectCredentialsAndInvoke(ofType(), ofType<ProxyDescribeEndpointGroup>()) } throws(EndpointGroupNotFoundException("NOT FOUND"))

        val request = ResourceHandlerRequest.builder<ResourceModel>().desiredResourceState(model).build()
        val response = DeleteHandler().handleRequest(proxy, request, null, logger)

        assertNotNull(response)
        assertEquals(OperationStatus.FAILED, response.status)
        assertEquals(0, response.callbackDelaySeconds)
        assertNull(response.callbackContext)
        assertEquals("Endpoint Group not found.", response.message)
        assertNull(response.resourceModels)
        assertEquals(model, response.resourceModel)
    }

    @Test
    fun handleRequest_ThresholdTimeExceeded() {
        val endpointConfigurations = ArrayList<EndpointConfiguration>()
        endpointConfigurations.add(EndpointConfiguration.builder().endpointId("EPID1").build())
        endpointConfigurations.add(EndpointConfiguration.builder().endpointId("EPID2").build())
        val model = ResourceModel.builder()
                .endpointGroupRegion(endpointGroupRegion)
                .listenerArn(listenerArn)
                .endpointGroupArn(endpointGroupArn)
                .healthCheckPort(20)
                .thresholdCount(3)
                .trafficDialPercentage(100.0)
                .healthCheckPath("/HEALTH")
                .endpointConfigurations(endpointConfigurations)
                .build()
        val request = ResourceHandlerRequest.builder<ResourceModel>().desiredResourceState(model).build()
        val callbackContext = CallbackContext(0, pendingStabilization = true)
        val handler = DeleteHandler()

        val exception = assertThrows(RuntimeException::class.java) {
            handler.handleRequest(proxy, request, callbackContext, logger)
        }

        assertEquals("Timed out waiting for endpoint group to be deployed.", exception.message)
    }

    @Test
    fun handleRequest_AcceleratorDoesntExist() {
        val endpointConfigurations = ArrayList<EndpointConfiguration>()
        endpointConfigurations.add(EndpointConfiguration.builder().endpointId("EPID1").build())
        endpointConfigurations.add(EndpointConfiguration.builder().endpointId("EPID2").build())
        val model = ResourceModel.builder()
                .endpointGroupRegion(endpointGroupRegion)
                .listenerArn(listenerArn)
                .endpointGroupArn(endpointGroupArn)
                .healthCheckPort(20)
                .thresholdCount(3)
                .trafficDialPercentage(100.0)
                .healthCheckPath("/HEALTH")
                .endpointConfigurations(endpointConfigurations)
                .build()

        every { proxy.injectCredentialsAndInvoke(ofType(), ofType<ProxyDescribeAccelerator>()) } throws(AcceleratorNotFoundException("NOT FOUND"))

        val request = ResourceHandlerRequest.builder<ResourceModel>().desiredResourceState(model).build()
        val callbackContext = CallbackContext(10, pendingStabilization = true)
        val response = DeleteHandler().handleRequest(proxy, request, callbackContext, logger)

        assertNotNull(response)
        assertEquals(OperationStatus.SUCCESS, response.status)
        assertNull(response.callbackContext)
        assertNull(response.resourceModel)
        assertNull(response.message)
        assertNull(response.errorCode)
        assertNull(response.resourceModels)
    }
}
